controller/src/main/java/io/pravega/controller/store/stream/tables/HistoryRecord.java

```diff
@@ -9,238 +9,173 @@
-
-        final int backToTop = BitConverter.readInt(table, offset + tailLengthOffset(partial.getSegments().size()));
-
-        assert backToTop == partial.length;
-
-        return new HistoryRecord(partial.epoch, partial.segments, eventTime, offset, false);
-    }
-
-    public byte[] toBytePartial() {
-        byte[] b = new byte[PARTIAL_FIELDS_FIXED_LENGTH + segments.size() * Integer.BYTES];
-        toBytePartial(b);
-        return b;
-    }
-
-    private void toBytePartial(byte[] b) {
-        // length
-        BitConverter.writeInt(b, lengthOffset(), length);
-        // epoch
-        BitConverter.writeInt(b, epochOffset(), epoch);
-        // segments
-        for (int i = 0; i < segments.size(); i++) {
-            BitConverter.writeInt(b, segmentOffset() + i * Integer.BYTES, segments.get(i));
-        }
-        // start offset
-        BitConverter.writeInt(b, offsetOffset(segments.size()), length - Long.BYTES - Integer.BYTES);
-    }
-
-    public byte[] remainingByteArray() {
-        byte[] b = new byte[REMAINING_FIELDS_FIXED_LENGTH];
-        remainingByteArray(b, 0);
-        return b;
-    }
-
-    private void remainingByteArray(byte[] b, int start) {
-        BitConverter.writeLong(b, start, scaleTime);
-        BitConverter.writeInt(b, start + Long.BYTES, length);
-    }
-
-    public byte[] toByteArray() {
-        byte[] b = new byte[FIXED_FIELDS_LENGTH + segments.size() * Integer.BYTES];
-        toBytePartial(b);
-        remainingByteArray(b, PARTIAL_FIELDS_FIXED_LENGTH + segments.size() * Integer.BYTES);
-        return b;
-    }
-
-    private static int getCount(int length) {
-        return (length - FIXED_FIELDS_LENGTH) / Integer.BYTES;
-    }
-
-    private static int lengthOffset() {
-        return 0;
-    }
-
-    private static int epochOffset() {
-        return lengthOffset() + Integer.BYTES;
-    }
-
-    private static int segmentOffset() {
-        return epochOffset() + Integer.BYTES;
-    }
-
-    private static int offsetOffset(int count) {
-        return segmentOffset() + count * Integer.BYTES;
-    }
-
-    private static int scaleTimeOffset(int segmentCount) {
-        return offsetOffset(segmentCount) + Integer.BYTES;
+    /**
+     * Method to fetch record immediately preceeding the given record.
+     * @param record record whose next record is to be found.
+     * @param historyTable history table
+     * @param historyIndex history index
+     * @return returns history record immediately preceeding given record. If the given record is first record
+     * in the table then return empty.
+     */
+    public static Optional<HistoryRecord> fetchPrevious(final HistoryRecord record, final byte[] historyIndex, final byte[] historyTable) {
+        return readRecord(record.epoch - 1, historyIndex, historyTable, true);
     }
 
-    private static int tailLengthOffset(int segmentCount) {
-        return scaleTimeOffset(segmentCount) + Long.BYTES;
+    @SneakyThrows
```

Comment:
`SneakyThrows(IOException.class)`

And really, it can throw a legitimate IOException at you, so I don't understand why you want to suppress its declaration.

Reply:
Its a checked exception that we will have to handle everywhere else.